git/gogit/git_client.go

```diff
@@ -0,0 +1,296 @@
+/*
+Copyright 2022 The Flux authors
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package gogit
+
+import (
+	"context"
+	"fmt"
+	"io"
+	"os"
+	"path/filepath"
+	"time"
+
+	"github.com/ProtonMail/go-crypto/openpgp"
+	extgogit "github.com/go-git/go-git/v5"
+	"github.com/go-git/go-git/v5/config"
+	"github.com/go-git/go-git/v5/plumbing"
+	"github.com/go-git/go-git/v5/plumbing/object"
+
+	"github.com/fluxcd/pkg/git"
+)
+
+// GoGitClient implements git.GitClient
+type GoGitClient struct {
+	path       string
+	repository *extgogit.Repository
+	authOpts   *git.AuthOptions
+}
+
+var _ git.GitClient = &GoGitClient{}
+
+// NewGoGitClient returns a new GoGitClient
+func NewGoGitClient(path string, authOpts *git.AuthOptions) *GoGitClient {
+	return &GoGitClient{
+		path:     path,
+		authOpts: authOpts,
+	}
+}
+
+func (g *GoGitClient) Init(ctx context.Context, url, branch string) error {
+	if g.repository != nil {
+		return nil
+	}
+
+	r, err := extgogit.PlainInit(g.path, false)
+	if err != nil {
+		return err
+	}
+
+	if _, err = r.CreateRemote(&config.RemoteConfig{
+		Name: extgogit.DefaultRemoteName,
+		URLs: []string{url},
+	}); err != nil {
+		return err
+	}
+
+	branchRef := plumbing.NewBranchReferenceName(branch)
+	if err = r.CreateBranch(&config.Branch{
+		Name:   branch,
+		Remote: extgogit.DefaultRemoteName,
+		Merge:  branchRef,
```

Comment:
The branch names are less relevant on the local repository (which is mostly ephemeral), than on the remote repository. But I suspect we want to be using the same branch name in both.